### ISSUE TITLE
pug template cache is never a 'hit'

### ISSUE THREAD

**Issue by moos** [sitespeedio/sitespeed.io#1218]
Due to the async nature of the render* calls, the template cache is never hit. 

```shell
$ bin/sitespeed.js -m3 -n1 http://0.0.0.0:8082
[2016-09-24 00:27:18] Versions OS: darwin 14.5.0 nodejs: v6.5.0 sitespeed.io: 4.0.0-beta.1 browsertime: 1.0.0-beta.3 coach: 0.25.0
[2016-09-24 00:27:19] Starting chrome for analysing http://0.0.0.0:8082 1 time(s)
[2016-09-24 00:27:19] Testing url http://0.0.0.0:8082 run 1
[2016-09-24 00:27:26] Starting chrome for analysing http://0.0.0.0:8082/ 1 time(s)
[2016-09-24 00:27:26] Testing url http://0.0.0.0:8082/ run 1
[2016-09-24 00:27:32] Starting chrome for analysing http://0.0.0.0:8082/.git/ 1 time(s)
[2016-09-24 00:27:32] Testing url http://0.0.0.0:8082/.git/ run 1
[2016-09-24 00:27:39] Render HTML for 3 page(s) 
assets.pug cached? false
toplist.pug cached? false
index.pug cached? false
detailed.pug cached? false
domains.pug cached? false
pages.pug cached? false
help.pug cached? false
url/index.pug cached? false
url/index.pug cached? false
url/index.pug cached? false
  compiled assets.pug 1
    render assets 1
  compiled index.pug 2
    render index 2
  compiled toplist.pug 3
    render toplist 3
  compiled detailed.pug 4
    render detailed 4
  compiled pages.pug 5
    render pages 5
  compiled domains.pug 6
    render domains 6
  compiled help.pug 7
    render help 7
  compiled url/index.pug 8
    render url/index 8
  compiled url/index.pug 9
    render url/index 9
  compiled url/index.pug 10
    render url/index 10
url/run.pug cached? false
url/run.pug cached? false
url/run.pug cached? false
  compiled url/run.pug 11
    render url/run 11
  compiled url/run.pug 12
    render url/run 12
  compiled url/run.pug 13
    render url/run 13
[2016-09-24 00:27:43] HTML stored in /Users/dev/git/sitespeed.io/sitespeed-result/0.0.0.0/2016-09-24-00-27-18
[2016-09-24 00:27:43] Finished analysing http://0.0.0.0:8082
```
Notice multiple calls on `url/index` and `url/run`.  There should be 9 (7 summary + 2 url) compile calls, but there are 13 = (7 summary + 2 url * 3 count).

**Comment by tobli** [comment]
Great find! Async is hard, for sure. Got any ideas for how to fix it? I'm thinking we should put a Promise for the compiled template in the cache around here: https://github.com/sitespeedio/sitespeed.io/blob/master/lib/plugins/html/renderer.js#L25 instead of waiting for the fs.readFileAsync to complete before populating the cache. Does that make sense?

We'd love it if you could share some details about the methods and tools you're using to debug this.
Thanks again!
/Tobias

**Comment by moos** [comment]
Tooling? No great mystery really -- just started inserting a few console.logs and immediately noticed the oddity.   May have a fix -- stay tuned -- need to clean things up a bit.

**Comment by moos** [comment]
Fix for #1219 seems to have fixed this as well.
```shell
$ bin/sitespeed.js -m3 -n1 http://0.0.0.0:8082
[2016-09-25 22:09:07] Versions OS: darwin 14.5.0 nodejs: v6.5.0 sitespeed.io: 4.0.0-beta.2 browsertime: 1.0.0-beta.5 coach: 0.27.0
[2016-09-25 22:09:08] Starting chrome for analysing http://0.0.0.0:8082 1 time(s)
[2016-09-25 22:09:08] Testing url http://0.0.0.0:8082 run 1
[2016-09-25 22:09:13] Starting chrome for analysing http://0.0.0.0:8082/ 1 time(s)
[2016-09-25 22:09:13] Testing url http://0.0.0.0:8082/ run 1
[2016-09-25 22:09:18] Starting chrome for analysing http://0.0.0.0:8082/.git/ 1 time(s)
[2016-09-25 22:09:18] Testing url http://0.0.0.0:8082/.git/ run 1
[2016-09-25 22:09:23] Render HTML for 3 page(s) 
assets.pug cached? false
  compiled assets.pug 1
    render assets 1
toplist.pug cached? false
  compiled toplist.pug 2
    render toplist 2
index.pug cached? false
  compiled index.pug 3
    render index 3
detailed.pug cached? false
  compiled detailed.pug 4
    render detailed 4
domains.pug cached? false
  compiled domains.pug 5
    render domains 5
pages.pug cached? false
  compiled pages.pug 6
    render pages 6
help.pug cached? false
  compiled help.pug 7
    render help 7
url/index.pug cached? false
  compiled url/index.pug 8
    render url/index 8
url/index.pug cached? true
    render url/index 9
url/index.pug cached? true
    render url/index 10
url/run.pug cached? false
  compiled url/run.pug 9
    render url/run 11
url/run.pug cached? true
    render url/run 12
url/run.pug cached? true
    render url/run 13
[2016-09-25 22:09:25] HTML stored in /Users/dev/git/sitespeed.io/sitespeed-result
[2016-09-25 22:09:25] Finished analysing http://0.0.0.0:8082
```
Only 9 compiles are run  - the cache is hit the other 4 times.

**Comment by soulgalore** [comment]
I've merged @moos fix for now, then later @tobli when we have time, we can make another shot to make it async. Thanks a lot again @moos for taking the time to fix it!
